### ISSUE TITLE
Add semicolons automatically

### ISSUE THREAD

**Issue by HaoDaWang** [rakudo/rakudo#2861]
Hi! I use rakudo recently, but I find it troublesome to add a semicolon at end of a statement, so can you provide semicolon automatically mechanism reference JS ASI? I think it's not difficult, so far, rakudo know where the semicolon is  missing.

**Comment by lizmat** [comment]
I think we're passed the stage where such a change would make it to the core language.  Having said that, it is probably technically possible to provide a Slang that would do just that in a  Perl 6 module.

**Comment by HaoDaWang** [comment]
That's awesome! Will it appear in the next version？

**Comment by lizmat** [comment]
Probably not.  First, someone will have to write a Perl 6 module to do that.  You will either have to be that person yourself, or find a person to do that for you.  Once it is ready for production, then maybe it could be considered for inclusion in the core.

**Comment by AlexDaniel** [comment]
Problems that require changing the language should be filed here: https://github.com/perl6/problem-solving

**Comment by HaoDaWang** [comment]
Oh, I'm sorry, should I create a new Issue again in that place?

**Comment by AlexDaniel** [comment]
I'd say yes. No matter what the final decision is, it'd be nice to have a written justification.

**Comment by HaoDaWang** [comment]
Thx for your reply, I will create a new Issue tomorrow.

**Comment by jnthn** [comment]
Rakudo is an implementation of the Perl 6 language, so this is really a language change request rather than simply a compiler change request. These days, those are handled through the [problem solving](https://github.com/perl6/problem-solving/issues) issue tracker, though given it'd just be assigned to me there anyway, I'll answer here. :)

While it might sort of look like Rakudo "knows" where the semicolon is missing, in reality it is [actually a very primitive heuristic catching a small number of common cases](https://github.com/rakudo/rakudo/blob/master/src/Perl6/Grammar.nqp#L981), which is very different from a general rule for implicit statement termination. The code location I linked to is a place where such a change could be inserted, and of course there's nothing to stop somebody developing a slang that overrides the `eat_terminator` rule by doing something different from the two error paths. Then it would be possible to write in that dialect of Perl 6 with different statement termination rules by sticking a `use` statement at the top of the program using the slang module. (Note that this may not be the only place that would need changes for a useful module doing such a language tweak.)

I don't think it would be fitting to include this in standard Perl 6, however, Some considerations as to why:

* There already is one implicit statement termination mechanism in Perl 6 (a `}` as the last non-whitespace thing on the line terminates the statement). People *do* trip over this one occasionally, and it seems likely that any further generalization of this mechanism would he liable to turn "occasionally" into "regularly".
* We put a lot of effort into error reporting, and a lot of the Perl 6 error reporting is derived from analyzing "two terms in a row" style errors (Perl 6's syntactic design leans very heavily on parsing term, infix, term, infix, term etc.. with few exceptions; the explicit statement terminator marks a point where we reset to looking for a term again). An implicit statement termination mechanism seems liable to therefore make other error reporting worse by turning these cases into implicit statement terminations, which then later result in other errors more distant from the original mistake.
* Since there are many cases of symbols being valid as both an infix or a term (especially since a term can start with a prefix), there's also a decent risk of such a mechanism turning what would today be wrong programs into syntactically valid but semantically bogus programs - in the worst case, ones that survive until runtime before failing.
* The [rules for JavaScript ASI](http://es5.github.io/#x7.9) are non-trivial, have well-known gotchas, are avoided by many JS developers as a result, and - perhaps their greatest sin - make gotchas for developers who choose to write the statement separators anyway. It's hard to imagine we'd not see the same set of problems with such a mechanism in Perl 6.

Perl's design does look to natural language. While use of a statement terminator is highly widespread in written natural languages, it's certainly the case that there are languages with no statement terminator character (using whitespace for it instead, I believe), so I'm sympathetic to the idea that some folks writing Perl 6 programs could find having to insert `;`s troublesome. However, I think the rest of the language design - especially error detection - draws too heavily on explicit marking of statement termination to make it optional in standard Perl 6. It's of course possible to design a language that doesn't require explicit statement terminators and get a good result, but I think you probably have to have it as a consideration from the start, and make other decisions differently in light of it. Most probably, too many choices have been made in Perl 6 predicted on having the `;` statement terminator in order to make it optional and not lose anything.
